ZEngine/src/CommandBuffer.cpp

```diff
@@ -0,0 +1,106 @@
+#include <pch.h>
+#include <Rendering/Buffers/CommandBuffer.h>
+#include <Rendering/ResourceTypes.h>
+#include <Hardwares/VulkanDevice.h>
+
+namespace ZEngine::Rendering::Buffers
+{
+    CommandBuffer::CommandBuffer(VkCommandPool command_pool, Rendering::QueueType type, bool present_on_swapchain)
+    {
+        m_queue_type = type;
+        if (present_on_swapchain)
+        {
+            m_signal_semaphore = CreateRef<Primitives::Semaphore>();
```

Comment:
As discussed, for visibility: you technically only need a semaphore per submit, not per command buffer. 

Keep in mind, you need a semaphore if you want a coarse-granular wait of one action on a queue to another action on a queue. So you could also have semaphores to synchronize submit-to-submit, not only submits-to-present. So the present_on_swapchain guard might not make sense in the future.